### ISSUE TITLE
[Bug]: slot cannot be passed a data-* attribute

### ISSUE THREAD

**Issue by george-cz** [microsoft/fluentui#27302]
### Library

React Components / v9 (@fluentui/react-components)

### System Info

```shell
N/A
```

### Are you reporting Accessibility issue?

no

### Reproduction

https://codesandbox.io/s/nice-poitras-wsk2c1?file=/example.tsx

## Actual Behavior
Setting a "data-something" attribute to a slot results in TS error.
The data attribute _is_ actually set on the appropriate DOM element.

## Expected Behavior
Data attributes should be allowed in slot.

### Requested priority

High

### Are you willing to submit a PR to fix?

yes

### Validations

- [X] Check that there isn't already an issue that reports the same bug to avoid creating a duplicate.
- [X] The provided reproduction is a minimal reproducible example of the bug.

**Comment by bsunderhus** [comment]
> @layershifter possibly, seems that the #23033 is about the component prop itself, what this one is about is passing down a data attribute to one of the slots of the component.
> 
> This might be caused by the same root issue, but I didn't want to make assumptions.

Yes, this basically the same issue, and as @behowell described, this would be a possible solution https://github.com/microsoft/fluentui/issues/23033#issuecomment-1234931827.

Just to be clear, this is not a slot problem itself, it's a TS problem; In the case down  hese TS is not being forgiving when passing properties that don't belong to a specified type, although it allows on the jsx root itself and also on non overlapping props.

https://www.typescriptlang.org/play?#code/JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcARFDvmQNwBQdwAdjFlAXlnAEICuMMCEwAKxMOgDedOHFQAbCDACMALjgS4YMWqVwAvvRnzFAJjUSDdPQ1xDU8PgKFwAvHAAUWiOLWPBIsVQASlcAPnVpOEoYXigmOCZeOTl6SzosAA9IWDhbJns4ABEsDiT4N3cQl3D3SIAePyFImQB6FrgYAAtgdB64CABrFDkuiF4Ac06OgE8wLHQAEwgExX6ANzY5ZDBmuAXkGGQAWgyXMmA8o4Ajfn8yXbaO7vQOYDl0K6xcZF5ULhhZvM9kIAOTwCAbKBbMAAfjg0xwnQANHAQMgBkCmMs-vksLtjMoXBINGR9ocTmQ1OdLjcnEwyCivGAdPprDJWu0un0+oNhqMJlMAXNFssseDIdD8QoYCYiSSyccMpTyBchNdbkIyKzIqFdgBRDLIcByPEyOotRpMXVBehAA


To give a simple solution for the problem in hand, just cast the type and things will work! https://codesandbox.io/s/reverent-tu-f49l7i?file=/example.tsx

**Comment by miroslavstastny** [comment]
An option is to have mutliple .d.ts versions. We would need a build support for that, worth investigating.

**Comment by bsunderhus** [comment]
> An option is to have mutliple .d.ts versions. We would need a build support for that, worth investigating.

It wouldn't be entirely possible in our current TS version (v3.9). we would need a set of known `data-` attributes and include them as optional in the slot definition to avoid this to happen. I believe that for now (meanwhile we still don't support  TS 4.4) the best solution is to cast 😟

**Comment by miroslavstastny** [comment]
This issue is faced by multiple major partners and needs more attention. We need to recheck TS version matrix and discuss the timelines with @Hotell.

**Comment by bsunderhus** [comment]
Seems like our current TS version does support String Template literal type! But we've encountered another problem while investigating this through our partners: 

> error TS2590: Expression produces a union type that is too complex to represent.

Seems like a universal solution with Template literal types will not resolve this issue

**Comment by bsunderhus** [comment]
This problem sums up on two TS features:

1. [Excess Property Check](https://www.typescriptlang.org/docs/handbook/2/objects.html#excess-property-checks)
    * https://github.com/Microsoft/TypeScript/issues/12936
3. [Attribute type checking](https://www.typescriptlang.org/docs/handbook/jsx.html#attribute-type-checking)

We would need some sort of feature on TS side to disable EPC (Excess property checks) on object literals passed to slots. This is not supported by language! every object literal declaration is **always** EPC!

**Comment by DanielRosenwasser** [comment]
I opened up https://github.com/microsoft/TypeScript/issues/58898, but in the meantime one way to address this is to use this type instead of `Pick`:

```ts
type DistPick<T, K> =
  T extends infer _ ? (K extends keyof T ? Pick<T, K> : never) : never;
```

I *believe* the problem is that `Pick` requires calculating the type `keyof T`. In this case, `T` is a union, where each constituent has *many* members that must be reduced away when finding the effective set of keys. Something about the index signature is causing us to do a lot of work (probably creating intersections of each key with the template string type), and I haven't investigated it.

But to avoid all of that work, `DistPick` simply takes a key (which may or may not be constrained to the key set of all constituents) and maps a `Pick` over each type. That will typically create *much* smaller types than the original inputs.

The subtlety is that if `id` has a type that differs, you will end up with a union instead of a flat type like `Pick` would've given you ([see an example here](https://www.typescriptlang.org/play/?ts=5.4.5#code/C4TwDgpgBAglC8UDeAoKUDWAuKBnYATgJYB2A5gNwoC+KKokUAQgsmpjiQK4C2ARhAJVa9cNAAiRfAAUiAYwwAeACoAaKAGkAfAnbKoEAB7AIJACa4opAGaCoAVSgB+KAAoNB46YuYIIAPbWDs5Qsgoq6tpQnBAAboIAlNHsJHGCVKKMAGoAjKxhSnAAPszqAEQYZVooAPQ16FAAek50ADYQwFCxOTi5rEgcUACyAIbAABYAdAQj5v48rkkuZeMQra3+ZdFQOQBMAMxQ1BRQdVAA7v4EGLh0DNBZu6ySMvKFUCVM5ZXVZ+jNbQ6XV2vSeiAG2GGYymMzmCyWUBWaw2Wxwe0Ox1O9TM-gglhI-k6l2uQA)). So if that's a problem, you can wrap `DistPick` in a final `Pick`.

```ts
type DistPick<T, K> = Pick<
  T extends infer _ ? (K extends keyof T ? Pick<T, K> : never) : never,
  K,
>;
```

**Comment by layershifter** [comment]
bot keep open

**Comment by layershifter** [comment]
bot keep open
